_posts/2015-12-01-Property-based-testing-FsCheck-Resharper-NCrunch-Visual-Studio-2013.md

```diff
@@ -0,0 +1,158 @@
+                static member reverseIsAsTheOriginal (xs:int[]) =          
+                    Array.rev xs = xs
+    ```
+7. Add the following test to verify all the properties defined in the ```ListProperties``` type
+    ``` fs
+            [<Test>] 
+            let verifyAll () =      
+                Check.QuickThrowOnFailureAll<ListProperties>()
+    ```
+8. 	Enable NCrunch following *NCRUNCH > Enable NCrunch* from the main menu and let the engine catch the error
+9. If everything is in place you should see the visual error for the feedback
+	<center><img src="/assets/img/custom/blog/2015-12-01-Property-based-testing/NCrunch-visual-feed-back-error.png"></center>
+10. Run All Tests in the Resharper Runner to check the output
+	<center><img src="/assets/img/custom/blog/2015-12-01-Property-based-testing/Resharper-tests-run-error.png"></center>
+11. You should read a message similar to
+    ```
+        -- Checking ListProperties ---
+        System.Reflection.TargetInvocationException : Exception has been thrown by the target of an invocation.
+        ---> System.Exception : ListProperties.reverseIsAsTheOriginal-Falsifiable, after 2 tests (4 shrinks) (StdGen (163791131,296083933)):
+        Original:
+		[|2; -1; 1; 1|]
+        Shrunk:
+        [|1; 0|]
+    ```
+12. This shows that the property was falsifiable (as expected) but let's try to gather more informations to address the issue
+13. Change the test so that it provides a verbose output when failing
+    ```
+            [<Test>] 
+    		let verifyAll () = 
+                Check.VerboseThrowOnFailureAll<ListProperties>()
+    ```
+14. Re-run the tests through Resharper Runner and read the output
+    ```
+        0:
+		[|-1|]
+        1:
+		[|0|]
+		2:
+		[|1; -3; 0; 0|]
+		shrink:
+		[|-3; 0; 0|]
+		shrink:
+		[|-3; 0|]
+		shrink:
+		[|3; 0|]
+		shrink:
+		[|2; 0|]
+		shrink:
+		[|1; 0|]
+		System.Reflection.TargetInvocationException : Exception has been thrown by the target of an invocation.
+        ----> System.Exception : ListProperties.reverseIsAsTheOriginal-Falsifiable, after 3 tests (5 shrinks) (StdGen (939147306,296083935)):
+		Original:
+		[|1; -3; 0; 0|]
+		Shrunk:
+        [|1; 0|]
+    ```
+15. The output shows the "shrinking process" from the third input original value ```[|1; -3; 0; 0|]``` to the  "shrunk" input ```[|1; 0|]``` that still falsifies the property 
+16. Let's correct the property to make it not falsifiable
+    ``` fs
+            type ListProperties =  
+                static member reverseOfReverseIsAsTheOriginal (xs:int[]) = 
+                    Array.rev (Array.rev xs) = xs
+    ```
+17. Re-run the tests that should be all green by now and read the output. You'll notice it's very verbose as expected and shows all the inputs used to verify the property.
+
+### Conclusion
+You are now ready to experiment with FsCheck!
+Let me know what is your experience with property based testing in F# writing at alex@codurance.com.
```

Comment:
instead of an email, how about twitter handle - everyone can see and get involved in the conversation ;)